core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -1,187 +1,337 @@
 use {
-    itertools::Itertools,
+    crate::replay_stage::DUPLICATE_THRESHOLD,
     solana_gossip::{
         cluster_info::ClusterInfo, contact_info::ContactInfo, crds::Cursor, epoch_slots::EpochSlots,
     },
-    solana_runtime::{bank::Bank, epoch_stakes::NodeIdToVoteAccounts},
-    solana_sdk::{
-        clock::{Slot, DEFAULT_SLOTS_PER_EPOCH},
-        pubkey::Pubkey,
-        timing::AtomicInterval,
-    },
+    solana_sdk::{clock::Slot, pubkey::Pubkey, timing::AtomicInterval},
     std::{
-        collections::{BTreeMap, HashMap},
-        sync::{Arc, Mutex, RwLock},
+        collections::{HashMap, VecDeque},
+        hash::RandomState,
+        ops::Range,
+        sync::{
+            atomic::{AtomicU64, Ordering},
+            Arc, Mutex, RwLock,
+        },
     },
 };
 
 // Limit the size of cluster-slots map in case
 // of receiving bogus epoch slots values.
-const CLUSTER_SLOTS_TRIM_SIZE: usize = 524_288; // 512K
+// This also constraints the size of the datastructure
+// if we are really really far behind.
+const CLUSTER_SLOTS_TRIM_SIZE: usize = 50000;
+
```

Comment:
Dropping datastructure max size to something more reasonable. 512K slots at 2000 validators would occupy 40 GB of RAM. 50K is still huge but far more manageable in practice.